README
libelule = Libelule::Client.new('hostname', 'username','password')
libelule.one_way_sync_from_local_to_remote('target_dir','base dir')



CHANGELOG
# CHANGELOG

## v0.0.1

* First release

## v0.0.2

* Change name to libelule, because ftp_sync is already taken


test/test_helper.rb
$LOAD_PATH.unshift( File.join( File.dirname(__FILE__), "..", "lib") )

require "rubygems"
require 'test/unit'
require 'mocha'
require 'shoulda-context'
require 'libelule'


test/ftp_sync_test.rb
# encoding: UTF-8
require 'test_helper'

class FtpSync::ClientTest < Test::Unit::TestCase

  FTP_SOURCE_DIR = File.join("/tmp","ftp_sync_test")
  FTP_TARGET_DIR = "remote_assets"

  def ftp_string_for_filename(file_name)
    "-rw-r--r--    1 james     staff            6 Jan 07 03:54 #{file_name}"
  end

  def ftp_string_for_directory(directory_name)
    "drw-r--r--    1 james     staff            6 Jan 07 03:54 #{directory_name}"
  end

  FtpSync::Client.silence_log_messages = true

  context "one_way_sync_from_local_to_remote" do
    setup do
      FileUtils.mkdir_p(FTP_SOURCE_DIR)

      @ftp_client_test_double = stub_everything('fake_ftp_client')
      @ftp_fake_client_class = stub("FakeFTPClient", :new => @ftp_client_test_double)

      @ftp = FtpSync::Client.new("ftp.test.host", "ftpuser", "ftppassword")
      @ftp.ftp_client = @ftp_fake_client_class
    end

    teardown do
      FileUtils.rm_rf(FTP_SOURCE_DIR)
      FileUtils.rm_rf(FTP_TARGET_DIR)
    end

    should "open a connection to the host" do
      @ftp_fake_client_class.expects(:new).with('ftp.test.host').returns(@ftp_client_test_double)
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should "login into the ftp server" do
      @ftp_client_test_double.expects(:login).with('ftpuser','ftppassword')
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should 'copy new files to the remote host' do
      FileUtils.touch(File.join(FTP_SOURCE_DIR, 'test.txt'))
      @ftp_client_test_double.expects(:put).with("test.txt")
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should 'copy new directories' do
      FileUtils.mkdir(File.join(FTP_SOURCE_DIR, 'testdirectory'))
      @ftp_client_test_double.expects(:mkdir).with('testdirectory')
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should 'copy files of subdirectories' do
      FileUtils.mkdir(File.join(FTP_SOURCE_DIR, 'testdirectory'))
      FileUtils.touch File.join(FTP_SOURCE_DIR, 'testdirectory', 'subfile.txt')
      @ftp_client_test_double.expects(:put).with('subfile.txt')
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should 'not copy a file to the remote host if it exists already' do
      FileUtils.touch(File.join(FTP_SOURCE_DIR, 'test.txt'))
      @ftp_client_test_double.stubs(:ls).yields(ftp_string_for_filename("test.txt"))
      @ftp_client_test_double.expects(:put).with('test.txt').times(0)
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end
  end
  
  context 'clean up resources with a manifest' do
    setup do
      @ftp_client_test_double = stub_everything('fake_ftp_client')
      @ftp_fake_client_class = stub("FakeFTPClient", :new => @ftp_client_test_double)

      @ftp = FtpSync::Client.new("ftp.test.host", "ftpuser", "ftppassword")
      @ftp.ftp_client = @ftp_fake_client_class

      @ftp_client_test_double.stubs(:ls).multiple_yields(
        ftp_string_for_filename("fake.3219C.txt"),
        ftp_string_for_filename("fake.3211C.txt"),
        ftp_string_for_directory("subdir1")
      ).then.multiple_yields(
        ftp_string_for_filename("fake.2223C.txt"),
        ftp_string_for_filename("fake.2222C.txt"),
        ftp_string_for_directory("subdir2")
      ).then.multiple_yields(
        ftp_string_for_filename("fake.3333C.txt"),
        ftp_string_for_filename("fake.3332C.txt")
      )
      @ftp_client_test_double.stubs(:pwd).returns('/published', '/published', '/published/subdir1', '/published/subdir1', '/published/subdir1/subdir2')
      @manifest = ["/published/fake.3219C.txt", "/published/subdir1/fake.2223C.txt", "/published/subdir1/subdir2/fake.3333C.txt"]
    end
    
    should 'delete versions not in the manifest' do
      @ftp_client_test_double.expects(:delete).with('fake.3211C.txt')
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest)
    end
    
    should 'keep files in the manifest' do
      @ftp_client_test_double.expects(:delete).with('fake.3219C.txt').never
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest)      
    end
    
    should 'keep files in subdirectories' do
      @ftp_client_test_double.expects(:delete).with('fake.3219C.txt').never
      @ftp_client_test_double.expects(:delete).with('fake.2223C.txt').never
      @ftp_client_test_double.expects(:delete).with('fake.3333C.txt').never
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest)
    end
    
    should 'descend into subdirectories and delete' do
      @ftp_client_test_double.expects(:delete).with('fake.2222C.txt')
      @ftp_client_test_double.expects(:delete).with('fake.3332C.txt')
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest)            
    end
    
    should 'have a dry run option' do
      @ftp_client_test_double.expects(:delete).never
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest, :dry_run => true)
    end
  end

  context 'clean up resources by index' do
    setup do
      @ftp_client_test_double = stub_everything('fake_ftp_client')
      @ftp_fake_client_class = stub("FakeFTPClient", :new => @ftp_client_test_double)

      @ftp = FtpSync::Client.new("ftp.test.host", "ftpuser", "ftppassword")
      @ftp.ftp_client = @ftp_fake_client_class

      @ftp_client_test_double.stubs(:ls).multiple_yields(
        ftp_string_for_filename("flash.14695871.css"),
        ftp_string_for_filename("512K"),
        ftp_string_for_filename("test.txt"),
        ftp_string_for_filename("test.1000.txt"),
        ftp_string_for_filename("test.1001A.txt"),
        ftp_string_for_filename("test.1002B.txt"),
        ftp_string_for_filename("test.1003C.txt"),
        ftp_string_for_filename("test.1004C.txt")
        )
    end

    should 'delete old versions' do
      @ftp_client_test_double.expects(:delete).with('test.1000.txt')
      @ftp_client_test_double.expects(:delete).with('test.1001A.txt')
      @ftp_client_test_double.expects(:delete).with('test.1002B.txt')
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C')
    end

    should 'keep unrevisioned files' do
      @ftp_client_test_double.expects(:delete).with('test.txt').never
      @ftp_client_test_double.expects(:delete).with('512K').never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C')
    end
        
    should 'keep files with suffix C' do
      @ftp_client_test_double.expects(:delete).with('test.1004C.txt').never
      @ftp_client_test_double.expects(:delete).with('test.1003C.txt').never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C')
    end
    
    should 'keep files with a different versioning schema' do
      @ftp_client_test_double.expects(:delete).with('flash.14695871.css').never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C')
    end
    
    should 'make the suffix to keep configurable' do
      @ftp_client_test_double.expects(:delete).with('test.1001A.txt').never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'A')
    end
    
    should 'raise an exception if suffix to keep is blank' do
      assert_raises(ArgumentError){ @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => nil) }
    end
    
    should 'raise an exception if suffix is not a single uppercase letter' do
      assert_raises(ArgumentError){ @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'foo') }
    end
    
    should 'not delete files when the dry run option is set' do
      @ftp_client_test_double.expects(:delete).never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C', :dry_run => true)
    end
  end
end


test/libelule_test.rb
# encoding: UTF-8
require 'test_helper'

class Libelule::ClientTest < Test::Unit::TestCase

  FTP_SOURCE_DIR = File.join("/tmp","libelule_test")
  FTP_TARGET_DIR = "remote_assets"

  def ftp_string_for_filename(file_name)
    "-rw-r--r--    1 james     staff            6 Jan 07 03:54 #{file_name}"
  end

  def ftp_string_for_directory(directory_name)
    "drw-r--r--    1 james     staff            6 Jan 07 03:54 #{directory_name}"
  end

  Libelule::Client.silence_log_messages = true

  context "one_way_sync_from_local_to_remote" do
    setup do
      FileUtils.mkdir_p(FTP_SOURCE_DIR)

      @ftp_client_test_double = stub_everything('fake_ftp_client')
      @ftp_fake_client_class = stub("FakeFTPClient", :new => @ftp_client_test_double)

      @ftp = Libelule::Client.new("ftp.test.host", "ftpuser", "ftppassword")
      @ftp.ftp_client = @ftp_fake_client_class
    end

    teardown do
      FileUtils.rm_rf(FTP_SOURCE_DIR)
      FileUtils.rm_rf(FTP_TARGET_DIR)
    end

    should "open a connection to the host" do
      @ftp_fake_client_class.expects(:new).with('ftp.test.host').returns(@ftp_client_test_double)
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should "login into the ftp server" do
      @ftp_client_test_double.expects(:login).with('ftpuser','ftppassword')
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should 'copy new files to the remote host' do
      FileUtils.touch(File.join(FTP_SOURCE_DIR, 'test.txt'))
      @ftp_client_test_double.expects(:put).with("test.txt")
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should 'copy new directories' do
      FileUtils.mkdir(File.join(FTP_SOURCE_DIR, 'testdirectory'))
      @ftp_client_test_double.expects(:mkdir).with('testdirectory')
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should 'copy files of subdirectories' do
      FileUtils.mkdir(File.join(FTP_SOURCE_DIR, 'testdirectory'))
      FileUtils.touch File.join(FTP_SOURCE_DIR, 'testdirectory', 'subfile.txt')
      @ftp_client_test_double.expects(:put).with('subfile.txt')
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end

    should 'not copy a file to the remote host if it exists already' do
      FileUtils.touch(File.join(FTP_SOURCE_DIR, 'test.txt'))
      @ftp_client_test_double.stubs(:ls).yields(ftp_string_for_filename("test.txt"))
      @ftp_client_test_double.expects(:put).with('test.txt').times(0)
      @ftp.one_way_sync_from_local_to_remote(FTP_SOURCE_DIR, FTP_TARGET_DIR)
    end
  end
  
  context 'clean up resources with a manifest' do
    setup do
      @ftp_client_test_double = stub_everything('fake_ftp_client')
      @ftp_fake_client_class = stub("FakeFTPClient", :new => @ftp_client_test_double)

      @ftp = Libelule::Client.new("ftp.test.host", "ftpuser", "ftppassword")
      @ftp.ftp_client = @ftp_fake_client_class

      @ftp_client_test_double.stubs(:ls).multiple_yields(
        ftp_string_for_filename("fake.3219C.txt"),
        ftp_string_for_filename("fake.3211C.txt"),
        ftp_string_for_directory("subdir1")
      ).then.multiple_yields(
        ftp_string_for_filename("fake.2223C.txt"),
        ftp_string_for_filename("fake.2222C.txt"),
        ftp_string_for_directory("subdir2")
      ).then.multiple_yields(
        ftp_string_for_filename("fake.3333C.txt"),
        ftp_string_for_filename("fake.3332C.txt")
      )
      @ftp_client_test_double.stubs(:pwd).returns('/published', '/published', '/published/subdir1', '/published/subdir1', '/published/subdir1/subdir2')
      @manifest = ["/published/fake.3219C.txt", "/published/subdir1/fake.2223C.txt", "/published/subdir1/subdir2/fake.3333C.txt"]
    end
    
    should 'delete versions not in the manifest' do
      @ftp_client_test_double.expects(:delete).with('fake.3211C.txt')
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest)
    end
    
    should 'keep files in the manifest' do
      @ftp_client_test_double.expects(:delete).with('fake.3219C.txt').never
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest)      
    end
    
    should 'keep files in subdirectories' do
      @ftp_client_test_double.expects(:delete).with('fake.3219C.txt').never
      @ftp_client_test_double.expects(:delete).with('fake.2223C.txt').never
      @ftp_client_test_double.expects(:delete).with('fake.3333C.txt').never
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest)
    end
    
    should 'descend into subdirectories and delete' do
      @ftp_client_test_double.expects(:delete).with('fake.2222C.txt')
      @ftp_client_test_double.expects(:delete).with('fake.3332C.txt')
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest)            
    end
    
    should 'have a dry run option' do
      @ftp_client_test_double.expects(:delete).never
      @ftp.clean_up_with_manifest(FTP_TARGET_DIR, @manifest, :dry_run => true)
    end
  end

  context 'clean up resources by index' do
    setup do
      @ftp_client_test_double = stub_everything('fake_ftp_client')
      @ftp_fake_client_class = stub("FakeFTPClient", :new => @ftp_client_test_double)

      @ftp = Libelule::Client.new("ftp.test.host", "ftpuser", "ftppassword")
      @ftp.ftp_client = @ftp_fake_client_class

      @ftp_client_test_double.stubs(:ls).multiple_yields(
        ftp_string_for_filename("flash.14695871.css"),
        ftp_string_for_filename("512K"),
        ftp_string_for_filename("test.txt"),
        ftp_string_for_filename("test.1000.txt"),
        ftp_string_for_filename("test.1001A.txt"),
        ftp_string_for_filename("test.1002B.txt"),
        ftp_string_for_filename("test.1003C.txt"),
        ftp_string_for_filename("test.1004C.txt")
        )
    end

    should 'delete old versions' do
      @ftp_client_test_double.expects(:delete).with('test.1000.txt')
      @ftp_client_test_double.expects(:delete).with('test.1001A.txt')
      @ftp_client_test_double.expects(:delete).with('test.1002B.txt')
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C')
    end

    should 'keep unrevisioned files' do
      @ftp_client_test_double.expects(:delete).with('test.txt').never
      @ftp_client_test_double.expects(:delete).with('512K').never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C')
    end
        
    should 'keep files with suffix C' do
      @ftp_client_test_double.expects(:delete).with('test.1004C.txt').never
      @ftp_client_test_double.expects(:delete).with('test.1003C.txt').never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C')
    end
    
    should 'keep files with a different versioning schema' do
      @ftp_client_test_double.expects(:delete).with('flash.14695871.css').never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C')
    end
    
    should 'make the suffix to keep configurable' do
      @ftp_client_test_double.expects(:delete).with('test.1001A.txt').never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'A')
    end
    
    should 'raise an exception if suffix to keep is blank' do
      assert_raises(ArgumentError){ @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => nil) }
    end
    
    should 'raise an exception if suffix is not a single uppercase letter' do
      assert_raises(ArgumentError){ @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'foo') }
    end
    
    should 'not delete files when the dry run option is set' do
      @ftp_client_test_double.expects(:delete).never
      @ftp.clean_up(FTP_TARGET_DIR, :suffix_to_keep => 'C', :dry_run => true)
    end
  end
end


lib/ftp_sync.rb
# encoding: UTF-8
require "ftp_sync/version"
require 'net/ftp'
module FtpSync
  class Client
    class << self
      attr_accessor :silence_log_messages
      silence_log_messages = true
    end

    attr_accessor :ftp_client

    def initialize(host, user, password)
      @host = host
      @user = user
      @password = password
      @ftp_client = Net::FTP
    end

    def ftp
      if @ftp.nil? || @ftp.closed?
        @ftp = ftp_client.send(:new, @host)
        @ftp.passive = true
      end
      @ftp
    end

    def one_way_sync_from_local_to_remote(local_dir, remote_dir)
      begin
        ftp.login(@user, @password)
        log "logged in, start syncing..."
        sync_folder(local_dir, remote_dir)
        log "sync finished"
        return true
      rescue Net::FTPPermError => e
        log "Failed: #{e.message}"
        return false
      end
    end

    def clean_up(target_dir, options)
      suffix_to_keep = options[:suffix_to_keep]
      dry_run = options[:dry_run]
      raise ArgumentError, 'please provide a suffix to keep' unless /[A-Z]/ =~ suffix_to_keep
      ftp.login(@user, @password)
      clean_up_folder(target_dir, suffix_to_keep, dry_run)
    end

    def clean_up_with_manifest(target_dir, manifest, options = {})
      ftp.login(@user, @password)
      dry_run = options[:dry_run]
      clean_up_folder_with_manifest(target_dir, manifest, dry_run)
    end

    private

    def clean_up_folder_with_manifest(target_dir, manifest, dry_run = false)
      ftp.chdir(target_dir)
      remote_dirs, remote_files = get_remote_dir_and_file_names
      remote_files.each do |file|
        full_path = File.join(ftp.pwd, file)
        if manifest.include?(full_path)
          log "[KEEP] file #{full_path}"
        else
          ftp.delete(file) unless dry_run
          log "[DELETE] file #{full_path}"
        end
      end
      remote_dirs.each do |dir|
        clean_up_folder_with_manifest(dir, manifest, dry_run)
      end
      ftp.chdir("..")
    end

    def clean_up_folder(target_dir, suffix_to_keep, dry_run = false)
      ftp.chdir(target_dir)
      remote_dirs, remote_files = get_remote_dir_and_file_names
      remote_files.each do |file|
        if file =~ /\.[0-9]{1,5}[^#{suffix_to_keep}]?\./
          log "deleting file #{target_dir}/#{file}"
          ftp.delete(file) unless dry_run
        end
      end
      remote_dirs.each do |dir|
        clean_up_folder(dir, suffix_to_keep, dry_run)
      end
      ftp.chdir("..")
    end

    def put_title(title)
      log "#{'-'*80}\n#{title}:\n\n"
    end

    def full_file_path(file)
      File.join(Dir.pwd, file)
    end

    def upload_file(file)
      put_title "upload file: #{full_file_path(file)}"
      ftp.put(file)
    end

    def upload_folder(dir)
      put_title "upload folder: #{full_file_path(dir)}"
      Dir.chdir dir
      ftp.mkdir dir
      ftp.chdir dir

      local_dirs, local_files = get_local_dir_and_file_names

      local_dirs.each do |subdir|
        upload_folder(subdir)
      end

      local_files.each do |file|
        upload_file(file)
      end

      Dir.chdir("..")
      ftp.chdir("..")
    end

    def sync_folder(local_dir, remote_dir)
      Dir.chdir local_dir
      ftp.chdir remote_dir

      put_title "process folder: #{Dir.pwd}"

      local_dirs, local_files = get_local_dir_and_file_names
      remote_dirs, remote_files = get_remote_dir_and_file_names

      new_dirs = local_dirs - remote_dirs
      new_files = local_files - remote_files
      existing_dirs = local_dirs - new_dirs
      existing_files = local_files - new_files

      new_files.each do |file|
        upload_file(file)
      end

      new_dirs.each do |dir|
        upload_folder(dir)
      end

      existing_dirs.each do |dir|
        sync_folder(dir, dir)
      end

      Dir.chdir("..")
      ftp.chdir("..")
    end

    def get_local_dir_and_file_names
      dirs = []
      files = []
      Dir.glob("*").each do |file|
        if File.file?(file)
          files << file
        else
          dirs << file
        end
      end
      return [dirs, files]
    end

    def get_remote_dir_and_file_names
      dirs = []
      files = []
      ftp.ls do |file|
        #-rw-r--r--    1 james     staff            6 Jan 07 03:54 hello.txt
        file_name = file.gsub(/\S+\s+\d+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+/, '')
        case file[0, 1]
        when "-"
          files << file_name
        when "d"
          dirs << file_name
        end
      end
      return [dirs, files]
    end

    def log(message)
      puts message unless self.class.silence_log_messages
    end
  end
end


lib/libelule.rb
# encoding: UTF-8
require "libelule/version"
require 'net/ftp'
module Libelule
  class Client
    class << self
      attr_accessor :silence_log_messages
      silence_log_messages = true
    end

    attr_accessor :ftp_client

    def initialize(host, user, password)
      @host = host
      @user = user
      @password = password
      @ftp_client = Net::FTP
    end

    def ftp
      if @ftp.nil? || @ftp.closed?
        @ftp = ftp_client.send(:new, @host)
        @ftp.passive = true
      end
      @ftp
    end

    def one_way_sync_from_local_to_remote(local_dir, remote_dir)
      begin
        ftp.login(@user, @password)
        log "logged in, start syncing..."
        sync_folder(local_dir, remote_dir)
        log "sync finished"
        return true
      rescue Net::FTPPermError => e
        log "Failed: #{e.message}"
        return false
      end
    end

    def clean_up(target_dir, options)
      suffix_to_keep = options[:suffix_to_keep]
      dry_run = options[:dry_run]
      raise ArgumentError, 'please provide a suffix to keep' unless /[A-Z]/ =~ suffix_to_keep
      ftp.login(@user, @password)
      clean_up_folder(target_dir, suffix_to_keep, dry_run)
    end

    def clean_up_with_manifest(target_dir, manifest, options = {})
      ftp.login(@user, @password)
      dry_run = options[:dry_run]
      clean_up_folder_with_manifest(target_dir, manifest, dry_run)
    end

    private

    def clean_up_folder_with_manifest(target_dir, manifest, dry_run = false)
      ftp.chdir(target_dir)
      remote_dirs, remote_files = get_remote_dir_and_file_names
      remote_files.each do |file|
        full_path = File.join(ftp.pwd, file)
        if manifest.include?(full_path)
          log "[KEEP] file #{full_path}"
        else
          ftp.delete(file) unless dry_run
          log "[DELETE] file #{full_path}"
        end
      end
      remote_dirs.each do |dir|
        clean_up_folder_with_manifest(dir, manifest, dry_run)
      end
      ftp.chdir("..")
    end

    def clean_up_folder(target_dir, suffix_to_keep, dry_run = false)
      ftp.chdir(target_dir)
      remote_dirs, remote_files = get_remote_dir_and_file_names
      remote_files.each do |file|
        if file =~ /\.[0-9]{1,5}[^#{suffix_to_keep}]?\./
          log "deleting file #{target_dir}/#{file}"
          ftp.delete(file) unless dry_run
        end
      end
      remote_dirs.each do |dir|
        clean_up_folder(dir, suffix_to_keep, dry_run)
      end
      ftp.chdir("..")
    end

    def put_title(title)
      log "#{'-'*80}\n#{title}:\n\n"
    end

    def full_file_path(file)
      File.join(Dir.pwd, file)
    end

    def upload_file(file)
      put_title "upload file: #{full_file_path(file)}"
      ftp.put(file)
    end

    def upload_folder(dir)
      put_title "upload folder: #{full_file_path(dir)}"
      Dir.chdir dir
      ftp.mkdir dir
      ftp.chdir dir

      local_dirs, local_files = get_local_dir_and_file_names

      local_dirs.each do |subdir|
        upload_folder(subdir)
      end

      local_files.each do |file|
        upload_file(file)
      end

      Dir.chdir("..")
      ftp.chdir("..")
    end

    def sync_folder(local_dir, remote_dir)
      Dir.chdir local_dir
      ftp.chdir remote_dir

      put_title "process folder: #{Dir.pwd}"

      local_dirs, local_files = get_local_dir_and_file_names
      remote_dirs, remote_files = get_remote_dir_and_file_names

      new_dirs = local_dirs - remote_dirs
      new_files = local_files - remote_files
      existing_dirs = local_dirs - new_dirs
      existing_files = local_files - new_files

      new_files.each do |file|
        upload_file(file)
      end

      new_dirs.each do |dir|
        upload_folder(dir)
      end

      existing_dirs.each do |dir|
        sync_folder(dir, dir)
      end

      Dir.chdir("..")
      ftp.chdir("..")
    end

    def get_local_dir_and_file_names
      dirs = []
      files = []
      Dir.glob("*").each do |file|
        if File.file?(file)
          files << file
        else
          dirs << file
        end
      end
      return [dirs, files]
    end

    def get_remote_dir_and_file_names
      dirs = []
      files = []
      ftp.ls do |file|
        #-rw-r--r--    1 james     staff            6 Jan 07 03:54 hello.txt
        file_name = file.gsub(/\S+\s+\d+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+/, '')
        case file[0, 1]
        when "-"
          files << file_name
        when "d"
          dirs << file_name
        end
      end
      return [dirs, files]
    end

    def log(message)
      puts message unless self.class.silence_log_messages
    end
  end
end
